### DISCUSSION TITLE
Trying to plug NER model with Relation Extraction model in pipeline

### DISCUSSION THREAD

**Discussion by leot13** [explosion/spaCy#7433] [Help: Coding & Implementations] [answered]
Hi, 
I did the tutorials for a custom NER model and the custom relationship extractor. I'm currently trying to create a pipeline with both of them, but I get this error related to the corpora reader when I use "spacy.Corpora.v1" :
![image](https://user-images.githubusercontent.com/17809020/111076686-b7ad2900-84c3-11eb-9a0d-c19f372a33bc.png)

I also get this error when I use the custom one "Gold_ents_Corpus.v1" used in the relationship extraction tutorial (I just delete the part when it sets the pred.ents=gold.ents because I want ot predict them):
![image](https://user-images.githubusercontent.com/17809020/111076796-3d30d900-84c4-11eb-9483-7c07dba0693f.png)

Could you help me figure out why those readers become a problem when I try to tie those 2 models together? 
Here are the important parts of my config file in case that helps:

```
[paths]
train = null
dev = null
raw = null
init_tok2vec = null

[system]
seed = 342
gpu_allocator = "pytorch"

[nlp]
lang = "en"
pipeline = ["transformer","ner", "relation_extractor"]
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}
batch_size = 1000

[components]

[components.transformer]
factory = "transformer"
max_batch_items = 4096
set_extra_annotations = {"@annotation_setters":"spacy-transformers.null_annotation_setter.v1"}

[components.transformer.model]
@architectures = "spacy-transformers.TransformerModel.v1"
name = "roberta-base"
tokenizer_config = {"use_fast": true}

[components.transformer.model.get_spans]
@span_getters = "spacy-transformers.strided_spans.v1"
window = 64
stride = 48

[components.ner]
factory = "ner"

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = false
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0

[components.ner.model.tok2vec.pooling]
@layers = "reduce_mean.v1"

[components.relation_extractor]
factory = "relation_extractor"
threshold = 0.5

[components.relation_extractor.model]
@architectures = "rel_model.v1"

[components.relation_extractor.model.create_instance_tensor]
@architectures = "rel_instance_tensor.v1"

[components.relation_extractor.model.create_instance_tensor.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0

[components.relation_extractor.model.create_instance_tensor.tok2vec.pooling]
@layers = "reduce_mean.v1"

[components.relation_extractor.model.create_instance_tensor.pooling]
@layers = "reduce_mean.v1"

[components.relation_extractor.model.create_instance_tensor.get_instances]
@misc = "rel_instance_generator.v1"
max_length = 100

[components.relation_extractor.model.classification_layer]
@architectures = "rel_classification_layer.v1"
nI = null
nO = null

[initialize]

[initialize.components]

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
file = ${paths.dev}

[corpora.train]
@readers = "spacy.Corpus.v1"
file = ${paths.train}

[training]
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600000
max_epochs = 0
max_steps = 1000
eval_frequency = 100
frozen_components = []
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
before_to_disk = null
logger = {"@loggers":"spacy.ConsoleLogger.v1"}
```

**Comment by svlandeg** [comment, marked as answer]
Apologies that this is confusing! The original (for demo purposes) [`Gold_ents_Corpus.v1`](https://github.com/explosion/projects/blob/v3/tutorials/rel_component/scripts/custom_functions.py) uses the input parameter `file`:
```
@spacy.registry.readers("Gold_ents_Corpus.v1")
def create_docbin_reader(file: Path) -> Callable[["Language"], Iterable[Example]]:
    return partial(read_files, file)


def read_files(file: Path, nlp: "Language") -> Iterable[Example]:
```
so you can use it as 
```
[corpora.dev]
@readers = "Gold_ents_Corpus.v1"
file = ${paths.dev}
```

But spaCy's built-in [`spacy.Corpus.v1`](https://spacy.io/api/corpus) uses `path` as input, so you have to use it as:
```
[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
```

When you accidentally mix/match the parameter names - the config validation will say that it doesn't know what to do with the unknown parameter. So basically it always depends on the actual architecture (in this case the "reader") you're using, which other parameters you need to define in that same config block.

Let me know if you run into additional issues!

**Reply by leot13** [reply]
Thank you very much, that solved this problem! 
I'm having other issues, so if I can't figure them out myself, I'll post them here since it would still be under the same theme

**Reply by svlandeg** [reply]
Happy to hear that resolved it!

**Reply by leot13** [reply]
Hi,
I'm having another issue that I can't quite figure out yet. I think it has to do with the inputs fed to the custom component. 
I have changed the readers (spacy.Corpus.v1, Gold_ents_Corpus.v1 with pred.ents=gold.ents commented because I want to predict them, and even Gold_ents_Corpus.v1 with the pred.ents=gold.ents ), but I keep having this same message:
![full pipeline with gold_ents reader and pred=gol disabled 2021-03-16 155509](https://user-images.githubusercontent.com/17809020/111372888-9b4eef00-8671-11eb-8cb9-ff67ca7400f4.png)

The problem is obviously not a frozen component, but I guess there is somewhere in the custom code where I should pass on the NER predictions to the relation extractor. I'm just not sure where. 
The error message points to the evaluate.py, but since it does not even start the training, I'm not sure how evaluate.py could be the issue since it is supposed to be used later on.

Do you think you could help me with this?

**Reply by svlandeg** [reply]
It's not `evaluate.py` that is causing this, but rather the call to `nlp.evaluate` that is executed during the training loop, to evaluate the performance on the dev set (this is what generates the precision/recall/F numbers in the table printed to the console).

The function where this error is thrown then, is [`score_relations`](https://github.com/explosion/projects/blob/v3/tutorials/rel_component/scripts/rel_pipe.py#L204) in `rel_pipe`:
```
    for example in examples:
        gold = example.reference._.rel
        pred = example.predicted._.rel
        for key, pred_dict in pred.items():
            gold_labels = [k for (k, v) in gold[key].items() if v == 1.0]
```
In the demo, the REL extractor is using gold NER labels. This means that at time of evaluation, all the keys in the `_.rel` annotation dictionary will refer to existing entity spans of the gold document.

However, if you change this, you'll need to adjust the evaluation and decide what to do with relations that refer to entities that are not in gold. This means that the relation extractor is predicting relations on top of wrongly predicted entity spans. You have two options I think:
- Discard the relations that refer to wrong entity spans, as the NER's score is already taking those into account
- Consider these relations as false positives also for REL, as the relation shouldn't have been predicted

This will only influence the score you get for the REL component - there is no implication to the loss that is used internally, or the way the REL component is trained. Either choice still gives you the same REL model, but with a slightly differently reported accuracy metric.

**Reply by leot13** [reply]
Thank you! It took me a while to realize that 'k' and 'key' mean very diferent things so I was confused, but I fixed it. I went with discarding the relations that refer to wrong entity spans because it made more sense to me.
Everything seem to be working! I just need to fit it to my own data, but that shouldn't be too much trouble.

Thanks again for the help and also for your tutorial, it was extremely useful!

**Reply by svlandeg** [reply]
Happy to hear it! :-)

**Comment by svlandeg** [comment]
If you get this working and you have some code or numbers to show, feel free to post in the [Show & Tell section](https://github.com/explosion/spaCy/discussions/categories/show-tell)! It's always nice to hear about specific applications, what works, and what doesn't, and it could certainly inspire others :-)

**Comment by karndeepsingh** [comment]
@svlandeg I have trained NER model and Relational Extraction model separately But How can we integrate both of these two model and train together? Or we need to do training separately for NER and Relational Extraction. If so, then how I can integrate both model together.If you could just provide me the steps to follow It would be a great help. Thanks.

**Reply by polm** [reply]
Don't @ individual maintainers, it's disruptive. 

You can train NER and relation extraction at the same time, the config in the very first post in this thread is doing that. 

Your issue seems only vaguely related to this thread, so if you want help please open a new discussion.

**Reply by karndeepsingh** [reply]
@polm  Sure!

**Reply by karndeepsingh** [reply]
@polm  https://github.com/explosion/spaCy/discussions/9139  here is the new discussion I created please help. Thanks

**Reply by leot13** [reply]
Hey @karndeepsingh , if you're looking for the little trick that made it work for me it's in the rel_pipe script.

In the score relation method, you need to add an if statement to check if the key is part of the gold entities.


        `for key, pred_dict in pred.items():
              if key in gold:
                  gold_labels = [k for (k, v) in gold[key].items() if v == 1.0]
                  for k, v in pred_dict.items():
                      if v >= threshold:
                          if k in gold_labels:
                            micro_prf.tp += 1
                        else:
                            micro_prf.fp += 1
                    else:
                        if k in gold_labels:
                            micro_prf.fn += 1
    `

**Reply by karndeepsingh** [reply]
@leot13 Thanks alot. It worked. Just one more thing, how can I restart the training from last checkpoint in spacy?
